runtime/pkg/duckdbreplicator/backup.go

```diff
@@ -0,0 +1,335 @@
+	g.SetLimit(8)
+
+	objects := d.backup.List(&blob.ListOptions{
+		Delimiter: "/", // only list directories with a trailing slash and IsDir set to true
+	})
+
+	tblVersions := make(map[string]string)
+	for {
+		// Stop the loop if the ctx was cancelled
+		var stop bool
+		select {
+		case <-ctx.Done():
+			stop = true
+		default:
+			// don't break
+		}
+		if stop {
+			break // can't use break inside the select
+		}
+
+		obj, err := objects.Next(ctx)
+		if err != nil {
+			if errors.Is(err, io.EOF) {
+				break
+			}
+			return err
+		}
+		if !obj.IsDir {
+			continue
+		}
+
+		table := strings.TrimSuffix(obj.Key, "/")
+		d.logger.Debug("SyncWithObjectStorage: discovered table", slog.String("table", table))
+
+		// get version of the table
+		var backedUpVersion string
+		err = retry(func() error {
+			res, err := d.backup.ReadAll(ctx, filepath.Join(table, "version.txt"))
+			if err != nil {
+				return err
+			}
+			backedUpVersion = string(res)
+			return nil
+		})
+		if err != nil {
+			if gcerrors.Code(err) == gcerrors.NotFound {
+				// invalid table directory
+				d.logger.Debug("SyncWithObjectStorage: invalid table directory", slog.String("table", table))
+				_ = d.deleteBackup(ctx, table, "")
+			}
+			return err
+		}
+		tblVersions[table] = backedUpVersion
+
+		// check with current version
+		version, exists, _ := tableVersion(d.writePath, table)
+		if exists && version == backedUpVersion {
+			d.logger.Debug("SyncWithObjectStorage: table is already up to date", slog.String("table", table))
+			continue
+		}
+
+		tableDir := filepath.Join(d.writePath, table)
+		// truncate existing table directory
+		if err := os.RemoveAll(tableDir); err != nil {
+			return err
+		}
+		if err := os.MkdirAll(filepath.Join(tableDir, backedUpVersion), os.ModePerm); err != nil {
+			return err
+		}
+
+		tblIter := d.backup.List(&blob.ListOptions{Prefix: filepath.Join(table, backedUpVersion)})
+		// download all objects in the table and current version
+		for {
+			obj, err := tblIter.Next(ctx)
+			if err != nil {
+				if errors.Is(err, io.EOF) {
+					break
+				}
+				return err
+			}
+			g.Go(func() error {
+				return retry(func() error {
+					file, err := os.Create(filepath.Join(d.writePath, obj.Key))
+					if err != nil {
+						return err
+					}
+					defer file.Close()
+
+					rdr, err := d.backup.NewReader(ctx, obj.Key, nil)
+					if err != nil {
+						return err
+					}
+					defer rdr.Close()
+
+					_, err = io.Copy(file, rdr)
+					return err
+				})
+			})
+		}
+	}
+
+	// Wait for all outstanding downloads to complete
+	err := g.Wait()
+	if err != nil {
+		return err
+	}
+
+	// Update table versions
+	for table, version := range tblVersions {
+		err = os.WriteFile(filepath.Join(d.writePath, table, "version.txt"), []byte(version), fs.ModePerm)
+		if err != nil {
+			return err
+		}
+	}
+
+	// remove any tables that are not in backup
+	entries, err := os.ReadDir(d.writePath)
+	if err != nil {
+		return err
+	}
```

Comment:
It is not a problem right now because `syncWrite` is always called before a write handle has been created and as of now write and read files are duplicated.
Though we may need to rethink this based on your comments elsewhere that writes and reads should use same local files.